exla/lib/exla/mlir/value.ex

```diff
@@ -434,206 +542,414 @@ defmodule EXLA.MLIR.Value do
+
+  defp complex_part_type({:c, size}), do: {:f, div(size, 2)}
 
-    results = Enum.map(result_refs, &%Value{function: function, ref: &1})
+  defp op(%Function{} = function, op_name, operands, result_types, opts \\ []) do
+    opts = Keyword.validate!(opts, attributes: [], regions: [])
 
-    {results, %Region{ref: pred_ref}, %Region{ref: body_ref}}
+    %{ref: function_ref} = function
+
+    refs =
+      Enum.map(operands, fn
+        %Value{ref: ref, function: %Function{ref: ^function_ref}} -> ref
+      end)
+
+    refs =
+      EXLA.NIF.mlir_op(
+        function.ref,
+        op_name,
+        refs,
+        result_types,
+        opts[:attributes],
+        opts[:regions]
+      )
+      |> unwrap!()
+
+    Enum.map(refs, &%Value{function: function, ref: &1})
+  end
+
+  defp type_tensor(type, shape) do
+    shape_sequence = shape |> Tuple.to_list() |> Enum.map_join("", &"#{&1}x")
+    "tensor<#{shape_sequence}#{type_number(type)}>"
   end
 
-  def variadic_return(function, values) when is_list(values) do
-    refs = Enum.map(values, & &1.ref)
+  defp type_number({:pred, 8}), do: "i1"
+  defp type_number({:s, width}), do: "i#{width}"
+  defp type_number({:u, width}), do: "ui#{width}"
+  defp type_number({:f, width}), do: "f#{width}"
+  defp type_number({:bf, width}), do: "bf#{width}"
+  defp type_number({:c, 64}), do: "complex<f32>"
+  defp type_number({:c, 128}), do: "complex<f64>"
 
-    refs = EXLA.NIF.mlir_return(function.ref, refs) |> unwrap!()
+  defp type_token(), do: "!stablehlo.token"
 
-    Enum.map(refs, fn ref -> %Value{function: function, ref: ref} end)
+  defp type_tuple(children) do
+    "tuple<#{Enum.join(children, ", ")}>"
   end
 
-  def qr(%Value{function: function, ref: ref}, q_shape, r_shape)
-      when is_tuple(q_shape) and is_tuple(r_shape) do
-    {q_ref, r_ref} =
-      EXLA.NIF.mlir_qr(function.ref, ref, Tuple.to_list(q_shape), Tuple.to_list(r_shape))
-      |> unwrap!()
+  defp number_literal(value, type) do
+    cond do
+      Nx.Type.complex?(type) ->
+        {re, im} =
+          case value do
+            %Complex{re: re, im: im} -> {re, im}
+            true -> {1, 0}
+            false -> {0, 0}
+            n -> {n, 0}
+          end
+
+        subtype = complex_part_type(type)
+        "(#{number_literal(re, subtype)}, #{number_literal(im, subtype)})"
+
+      Nx.Type.float?(type) ->
+        hex_data = float_hex(value, type)
+        "0x#{hex_data}"
+
+      true ->
+        "#{value}"
+    end
+  end
 
-    {
-      %Value{function: function, ref: q_ref},
-      %Value{function: function, ref: r_ref}
-    }
+  defp float_hex(value, {_, size} = type) do
+    data =
+      case value do
+        :nan -> type |> Nx.Type.nan_binary() |> native_to_big()
+        :infinity -> type |> Nx.Type.infinity_binary() |> native_to_big()
+        :neg_infinity -> type |> Nx.Type.neg_infinity_binary() |> native_to_big()
+        value -> <<value::float-size(size)-big>>
+      end
+
+    Base.encode16(data)
   end
```

Comment:
Is MLIR necessarily always big endian?

Reply:
Would be nice to have a comment with a reference on that

Reply:
I didn't find any explicit reference, initially I tried BinaryBackend with `Nx.to_binary` and then realised it's big endian. The docs on [FloatAttr](https://mlir.llvm.org/docs/Dialects/Builtin/#floatattr) have an example for NaN and infinity, and they are big-endian, so I assume that's always the case.

> where the hexadecimal value is interpreted as bits of the underlying binary representation

Unless stated otherwise, I would interpret this as big-endian too.

But I will add a link to the docs anyway.